src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+        // totalAmountPledged_fromJB = paymentTerminalStore.balanceOf(paymentTerminal, projectId);
+        // if (totalAmountPledged != totalAmountPledged_fromJB) {
+        //     totalAmountPledged = totalAmountPledged_fromJB;
+        // Need fallback scenario here, or is ^ this adequate?
+        // }
+        if (totalAmountPledged >= cycleTarget) {
+            isTargetMet = true;
+        }
+
+        emit CycleHasClosed(isCycleExpired, isTargetMet);
+    }
+
+    /// @notice This function gets called when the project's token holders redeem.
+    /// @dev Part of IJBFundingCycleDataSource.
+    /// @param _data Standard Juicebox project redemption data. See https://docs.juicebox.money/dev/api/data-structures/jbredeemparamsdata/.
+    /// @return reclaimAmount Amount to be reclaimed from the treasury. This is useful for optionally customizing how much funds from the treasury are disbursed per redemption.
+    /// @return memo A memo to be forwarded to the event. Useful for describing any new actions are being taken.
+    /// @return delegateAllocations Amount to be sent to delegates instead of being added to the beneficiary. Useful for auto-routing funds from a treasury as redemptions are sought.
+    function redeemParams(JBRedeemParamsData calldata _data)
+        external
+        view
+        virtual
+        override
+        returns (
+            uint256 reclaimAmount,
+            string memory memo,
+            JBRedemptionDelegateAllocation3_1_1[] memory delegateAllocations
+        )
+    {
+        address payable redeemer = payable(_data.holder);
+
+        // The if statements are the only changes from the redeemParams() template.
+        if (pledgedAmount[redeemer] == 0) revert MustBePledger();
+        if (isCycleExpired != true) revert CycleHasNotEndedYet(cycleExpiryDate);
+        if (isTargetMet == true) revert NoRefundsForSuccessfulCycle();
+        if (hasBeenRefunded[redeemer] == true) revert AlreadyWithdrawnRefund();
+
+        // Forward the default reclaimAmount received from the protocol.
+        reclaimAmount = _data.reclaimAmount.value;
+        // Forward the default memo received from the redeemer.
+        memo = _data.memo;
+        // Add `this` contract as a Redeem Delegate so that it receives a `didRedeem` call. Don't send any extra funds to the delegate.
+        delegateAllocations = new JBRedemptionDelegateAllocation3_1_1[](1);
+        delegateAllocations[0] = JBRedemptionDelegateAllocation3_1_1(this, 0, "");
+
+        return (reclaimAmount, memo, delegateAllocations);
+    }
+
+    /// @notice If cycle meets project target, then this function cannot be called, since pauseRedemptions will be on.
+    /// If cycle fails to meet its target, then this function will proceed all the way through for pledgers.
+    function didRedeem(JBDidRedeemData3_1_1 calldata _data) external payable {
+        if (
+            msg.value != 0 || !directory.isTerminalOf(projectId, IJBPaymentTerminal(msg.sender))
```

Comment:
these `isTerminalOf` sender checks should be a modifier to clean up the code. Could also add one for msg.value == 0

Reply:
Good call, will do.